Automattic/vip-block-data-api
# Add GraphQL API

## Description

This PR is meant add a GraphQL API to the Block Data API. It assumes that you have WP-GraphQL installed, and uses a filter to tweak the output of the `ContentParser` to generate the GraphQL response.

In addition, it uses the `wp_unique_id` function to generate ids for the blocks so that the blocks can be correlated together. The `inner_blocks` are therefore able to be flattened entirely, thereby bypassing the nested depth problem in graphQL.

This is in the draft status to get initial feedback, as well as to add some tests and guard rails around the GraphQL API.

![graphql-rest-demo](https://github.com/Automattic/vip-block-data-api/assets/1771524/f6c90f82-0a47-4044-9ca6-a45ba3f2177a)

Resolves #49 

## Steps to Test

- Download, install and activate WP-GraphQL
- Activate Block Data API with this branch
- Modify the initial post to add some nested blocks
- Run the following GraphQL query
```graphQL
query NewQuery {
  post(id: "1", idType: DATABASE_ID) {
    blocksData {
      blocks {
        id
        name
        attributes {
          name
          value
        }
        innerBlocks {
          name
          parentId
          id
          attributes {
            name
            value
          }
        }
      }
    }
  }
}
```
- Ensure the output matches the post as well as the rest API
